From 963a43da5e2b8affefbad9140ebc92321e0342dd Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Mon, 19 Jul 2021 16:30:00 +0000
Subject: [PATCH backport v4.19 103/104] mlxsw: minimal: Add system event
 handler

Add system event handler for treating line card specific signals on
modular system. These signals indicate line card state changes, like
line card activation or de-activation.
When such signals are received, driver should create or destroy "hwmon"
and "thermal" objects, associated with line card in a slot, for which
signal has been received.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/net/ethernet/mellanox/mlxsw/minimal.c | 23 +++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/net/ethernet/mellanox/mlxsw/minimal.c b/drivers/net/ethernet/mellanox/mlxsw/minimal.c
index eab2a216c00c..cf2a4562abc8 100644
--- a/drivers/net/ethernet/mellanox/mlxsw/minimal.c
+++ b/drivers/net/ethernet/mellanox/mlxsw/minimal.c
@@ -306,6 +306,28 @@ static void mlxsw_m_ports_remove(struct mlxsw_m_area *mlxsw_m_area)
 	kfree(mlxsw_m_area->ports);
 }
 
+static void mlxsw_m_sys_event_handler(struct mlxsw_core *mlxsw_core)
+{
+	struct mlxsw_m *mlxsw_m = mlxsw_core_driver_priv(mlxsw_core);
+	struct mlxsw_linecards *linecards = mlxsw_core_linecards(mlxsw_core);
+	char mddq_pl[MLXSW_REG_MDDQ_LEN];
+	int i, err;
+
+	if (!mlxsw_m->linecards_registered || !linecards)
+		return;
+
+	/* Handle line cards, for which active status has been changed. */
+	for (i = 1; i <= linecards->count; i++) {
+		mlxsw_reg_mddq_slot_info_pack(mddq_pl, i, false);
+		err = mlxsw_reg_query(mlxsw_m->core, MLXSW_REG(mddq), mddq_pl);
+		if (err)
+			dev_err(mlxsw_m->bus_info->dev, "Fail to query MDDQ register for slot %d\n",
+				i);
+
+		mlxsw_linecard_status_process(mlxsw_m->core, mddq_pl);
+	}
+}
+
 static int mlxsw_m_init(struct mlxsw_core *mlxsw_core,
 			const struct mlxsw_bus_info *mlxsw_bus_info)
 {
@@ -347,6 +369,7 @@ static struct mlxsw_driver mlxsw_m_driver = {
 	.priv_size		= sizeof(struct mlxsw_m),
 	.init			= mlxsw_m_init,
 	.fini			= mlxsw_m_fini,
+	.sys_event_handler	= mlxsw_m_sys_event_handler,
 	.profile		= &mlxsw_m_config_profile,
 	.res_query_enabled	= true,
 };
-- 
2.20.1

